Remote debug over adb seems to be broken (2.1.4-beta)

**Operating system or device - Godot version:**
OSX, Android 4.4.2
Godot 2.1.4-beta (commit e5102a9)
**Issue description:**

If I set `Use remote debug over Adb` the debugging doesn't work at all.

[There I described the problem](https://godotengine.org/qa/17399/using-remote-debug-on-android-2-1-4-beta)

**Steps to reproduce:**

Set `Use remote debug over Adb` and try to debug on the android device.

**Link to minimal example project:**

@RandomShaper 
1) checked out commit 01ef7a73de46236c4b0cd0e7853f9d3c91768111 (current 2.1 HEAD)
2) build godot engine and android templates:

```
scons platform=osx target=release_debug tools=yes bits=64 -j12
export ANDROID_NDK_ROOT=/opt/android-ndk-r14b/
export ANDROID_HOME=/opt/android-sdk/
scons platform=android target=release android_stl=yes -j12
cd platform/android/java
./gradlew build
cd ../../../
scons platform=android target=release_debug android_stl=yes verbose=yes -j12 
cd platform/android/java
./gradlew build
cd ../../../
```
3) copy templates to ~/.godot/templates so engine can found them by default

4) adb version
```
bash-3.2$ adb version
Android Debug Bridge version 1.0.39
Revision 5943271ace17-android
```

5) device is android 4.4.2 (I will be able to check on 5+ or 6+ later)

6) Editor - Android - Force System User disabled

Deploy - Remote debug enabled

Deploy - Small deploy with Network FS enabled

7) Start project with green android button

![2017-08-30 16 18 54](https://user-images.githubusercontent.com/1177068/29874303-fe7e0676-8d9e-11e7-9e26-bd39560c9dd4.png)

8) catlog:

```
08-30 16:20:35.612: I/godot(5539): Remote Debugger: Connection failed with status: '3', retrying in 1 sec.
08-30 16:20:36.612: I/godot(5539): Remote Debugger: Connection failed with status: '3', retrying in 1 sec.
08-30 16:20:37.612: I/godot(5539): Remote Debugger: Connection failed with status: '3', retrying in 1 sec.
08-30 16:20:37.612: I/godot(5539): Remote Debugger: Unable to connect
08-30 16:20:37.616: E/godot(5539): Could not connect to remotefs: localhost:6010
08-30 16:20:37.616: I/godot(5539): Godot Engine v2.1.4.stable.custom_build (c) 2008-2017 Juan Linietsky, Ariel Manzur.
08-30 16:20:37.616: I/godot(5539): Usage: apk [options] [scene]
08-30 16:20:37.616: I/godot(5539): Options:
08-30 16:20:37.616: I/godot(5539): 	-path [dir] : Path to a game, containing engine.cfg
08-30 16:20:37.616: I/godot(5539): 	-test [test] : Run a test.
08-30 16:20:37.616: I/godot(5539): 		(
08-30 16:20:37.616: I/godot(5539): string
08-30 16:20:37.616: I/godot(5539): , containers
08-30 16:20:37.616: I/godot(5539): , math
08-30 16:20:37.616: I/godot(5539): , render
08-30 16:20:37.616: I/godot(5539): , particles
08-30 16:20:37.616: I/godot(5539): , multimesh
08-30 16:20:37.616: I/godot(5539): , gui
08-30 16:20:37.616: I/godot(5539): , io
08-30 16:20:37.616: I/godot(5539): , shaderlang
08-30 16:20:37.616: I/godot(5539): , physics
08-30 16:20:37.616: I/godot(5539): )
08-30 16:20:37.616: I/godot(5539): 	-r WIDTHxHEIGHT	 : Request Window Resolution
08-30 16:20:37.616: I/godot(5539): 	-p XxY	 : Request Window Position
08-30 16:20:37.616: I/godot(5539): 	-f		 : Request Fullscreen
08-30 16:20:37.616: I/godot(5539): 	-mx		 Request Maximized
08-30 16:20:37.616: I/godot(5539): 	-w		 Request Windowed
08-30 16:20:37.616: I/godot(5539): 	-vd DRIVER	 : Video Driver (
08-30 16:20:37.616: I/godot(5539): GLES2
08-30 16:20:37.616: I/godot(5539): )
08-30 16:20:37.616: I/godot(5539): 	-ldpi	 : Force low-dpi mode (OSX Only)
08-30 16:20:37.616: I/godot(5539): 	-ad DRIVER	 : Audio Driver (
08-30 16:20:37.616: I/godot(5539): Android
08-30 16:20:37.616: I/godot(5539): )
08-30 16:20:37.616: I/godot(5539): 	-rthread <mode>	 : Render Thread Mode ('unsafe', 'safe', 'separate').
08-30 16:20:37.616: I/godot(5539): 	-s,-script [script] : Run a script.
08-30 16:20:37.616: I/godot(5539): 	-d,-debug : Debug (local stdout debugger).
08-30 16:20:37.616: I/godot(5539): 	-rdebug ADDRESS : Remote debug (<ip>:<port> host address).
08-30 16:20:37.616: I/godot(5539): 	-fdelay [msec]: Simulate high CPU load (delay each frame by [msec]).
08-30 16:20:37.616: I/godot(5539): 	-timescale [msec]: Simulate high CPU load (delay each frame by [msec]).
08-30 16:20:37.616: I/godot(5539): 	-bp : breakpoint list as source::line comma separated pairs, no spaces (%20,%2C,etc instead).
08-30 16:20:37.616: I/godot(5539): 	-v : Verbose stdout mode
08-30 16:20:37.616: I/godot(5539): 	-lang [locale]: Use a specific locale
08-30 16:20:37.616: I/godot(5539): 	-rfs <host/ip>[:<port>] : Remote FileSystem.
08-30 16:20:37.616: I/godot(5539): 	-rfs_pass <password> : Password for Remote FileSystem.
08-30 16:20:37.630: I/godot(5539): [1;31mERROR: free_static: [0m[1mCondition ' !MemoryPoolStatic::get_singleton() ' is true.
08-30 16:20:37.631: I/godot(5539): [0;31m   At: core/os/memory.cpp:59.[0m
08-30 16:20:37.631: I/godot(5539): [1;31mERROR: free_static: [0m[1mCondition ' !MemoryPoolStatic::get_singleton() ' is true.
08-30 16:20:37.631: I/godot(5539): [0;31m   At: core/os/memory.cpp:59.[0m
08-30 16:20:37.631: I/godot(5539): [1;31mERROR: free_static: [0m[1mCondition ' !MemoryPoolStatic::get_singleton() ' is true.
08-30 16:20:37.631: I/godot(5539): [0;31m   At: core/os/memory.cpp:59.[0m
08-30 16:20:37.631: I/godot(5539): *****UNABLE TO SETUP
08-30 16:20:37.650: E/SensorManager(5539): sensor or listener is null
08-30 16:20:37.691: E/SensorManager(5539): sensor or listener is null
08-30 16:20:37.725: D/AndroidRuntime(5539): Shutting down VM
08-30 16:20:37.725: W/dalvikvm(5539): threadid=1: thread exiting with uncaught exception (group=0x419bac98)
08-30 16:20:37.725: W/dalvikvm(5539): threadid=1: uncaught exception occurred
08-30 16:20:37.726: W/System.err(5539): java.lang.RuntimeException: Unable to resume activity {org.asmstudio.spacecute/org.godotengine.godot.Godot}: java.lang.NullPointerException
08-30 16:20:37.733: W/System.err(5539): 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2964)
08-30 16:20:37.733: W/System.err(5539): 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2993)
08-30 16:20:37.734: W/System.err(5539): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2395)
08-30 16:20:37.734: W/System.err(5539): 	at android.app.ActivityThread.access$800(ActivityThread.java:151)
08-30 16:20:37.734: W/System.err(5539): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
08-30 16:20:37.734: W/System.err(5539): 	at android.os.Handler.dispatchMessage(Handler.java:110)
08-30 16:20:37.734: W/System.err(5539): 	at android.os.Looper.loop(Looper.java:193)
08-30 16:20:37.734: W/System.err(5539): 	at android.app.ActivityThread.main(ActivityThread.java:5292)
08-30 16:20:37.734: W/System.err(5539): 	at java.lang.reflect.Method.invokeNative(Native Method)
08-30 16:20:37.734: W/System.err(5539): 	at java.lang.reflect.Method.invoke(Method.java:515)
08-30 16:20:37.734: W/System.err(5539): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
08-30 16:20:37.734: W/System.err(5539): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
08-30 16:20:37.734: W/System.err(5539): 	at dalvik.system.NativeStart.main(Native Method)
08-30 16:20:37.735: W/System.err(5539): Caused by: java.lang.NullPointerException
08-30 16:20:37.735: W/System.err(5539): 	at org.godotengine.godot.Godot.onResume(Godot.java:630)
08-30 16:20:37.735: W/System.err(5539): 	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1197)
08-30 16:20:37.735: W/System.err(5539): 	at com.lbe.security.service.core.client.b.x.callActivityOnResume(Unknown Source)
08-30 16:20:37.735: W/System.err(5539): 	at android.app.Activity.performResume(Activity.java:5343)
08-30 16:20:37.735: W/System.err(5539): 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2950)
08-30 16:20:37.736: W/System.err(5539): 	... 12 more
08-30 16:20:37.736: W/dalvikvm(5539): threadid=1: calling UncaughtExceptionHandler
08-30 16:20:37.740: E/AndroidRuntime(5539): FATAL EXCEPTION: main
08-30 16:20:37.740: E/AndroidRuntime(5539): Process: org.asmstudio.spacecute, PID: 5539
08-30 16:20:37.740: E/AndroidRuntime(5539): java.lang.RuntimeException: Unable to resume activity {org.asmstudio.spacecute/org.godotengine.godot.Godot}: java.lang.NullPointerException
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2964)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2993)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2395)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread.access$800(ActivityThread.java:151)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.os.Handler.dispatchMessage(Handler.java:110)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.os.Looper.loop(Looper.java:193)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread.main(ActivityThread.java:5292)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at java.lang.reflect.Method.invokeNative(Native Method)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at java.lang.reflect.Method.invoke(Method.java:515)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at dalvik.system.NativeStart.main(Native Method)
08-30 16:20:37.740: E/AndroidRuntime(5539): Caused by: java.lang.NullPointerException
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at org.godotengine.godot.Godot.onResume(Godot.java:630)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1197)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at com.lbe.security.service.core.client.b.x.callActivityOnResume(Unknown Source)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.Activity.performResume(Activity.java:5343)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2950)
08-30 16:20:37.740: E/AndroidRuntime(5539): 	... 12 more
```

9) SAD

@RandomShaper 
I just checked this PR (https://github.com/godotengine/godot/pull/10792) and it works for godot test projects. At least I checked monitors, profiler and live scene updating. All of them works as expected.

But it always crashes on my own project:

```
08-30 22:18:30.718: W/dalvikvm(7467): Exception thrown (Ljava/lang/NoClassDefFoundError;) while throwing internal exception (Ljava/lang/OutOfMemoryError;)
08-30 22:18:30.718: I/godot(7467): [1;31mERROR: set_locale: [0m[1mInvalid Locale: 
08-30 22:18:30.718: I/godot(7467): [0;31m   At: core/translation.cpp:921.[0m
08-30 22:18:30.726: D/GraphicBuffer(7467): close handle(0x602b1668) (w:720 h:1280 f:1)
08-30 22:18:30.732: D/dalvikvm(7467): threadid=11: exiting
08-30 22:18:30.732: W/dalvikvm(7467): threadid=11: thread exiting with uncaught exception (group=0x419bac98)
08-30 22:18:30.732: W/dalvikvm(7467): threadid=11: uncaught exception occurred
08-30 22:18:30.733: W/System.err(7467): java.lang.NoClassDefFoundError: java/util/Locale
08-30 22:18:30.736: W/System.err(7467): 	at org.godotengine.godot.GodotIO.getLocale(GodotIO.java:516)
08-30 22:18:30.736: W/System.err(7467): 	at org.godotengine.godot.GodotLib.step(Native Method)
08-30 22:18:30.736: W/System.err(7467): 	at org.godotengine.godot.GodotView$Renderer.onDrawFrame(GodotView.java:618)
08-30 22:18:30.736: W/System.err(7467): 	at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1557)
08-30 22:18:30.736: W/System.err(7467): 	at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1263)
08-30 22:18:30.736: W/System.err(7467): Caused by: java.lang.ClassNotFoundException: Didn't find class "org/godotengine/godot/GodotAppodeal" on path: DexPathList[[zip file "/data/app/org.asmstudio.spacecute-1.apk"],nativeLibraryDirectories=[/data/app-lib/org.asmstudio.spacecute-1, /vendor/lib, /system/lib]]
08-30 22:18:30.739: W/System.err(7467): 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
08-30 22:18:30.739: W/System.err(7467): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:497)
08-30 22:18:30.739: W/System.err(7467): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
08-30 22:18:30.740: W/System.err(7467): 	... 4 more
08-30 22:18:30.740: W/dalvikvm(7467): threadid=11: calling UncaughtExceptionHandler
08-30 22:18:30.742: E/AndroidRuntime(7467): FATAL EXCEPTION: GLThread 36633
08-30 22:18:30.742: E/AndroidRuntime(7467): Process: org.asmstudio.spacecute, PID: 7467
08-30 22:18:30.742: E/AndroidRuntime(7467): java.lang.NoClassDefFoundError: java/util/Locale
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at org.godotengine.godot.GodotIO.getLocale(GodotIO.java:516)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at org.godotengine.godot.GodotLib.step(Native Method)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at org.godotengine.godot.GodotView$Renderer.onDrawFrame(GodotView.java:618)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1557)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1263)
08-30 22:18:30.742: E/AndroidRuntime(7467): Caused by: java.lang.ClassNotFoundException: Didn't find class "org/godotengine/godot/GodotAppodeal" on path: DexPathList[[zip file "/data/app/org.asmstudio.spacecute-1.apk"],nativeLibraryDirectories=[/data/app-lib/org.asmstudio.spacecute-1, /vendor/lib, /system/lib]]
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:497)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
08-30 22:18:30.742: E/AndroidRuntime(7467): 	... 4 more
```

--- a/platform/android/export/export.cpp
+++ b/platform/android/export/export.cpp
@@ -237,7 +237,7 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		String id;
 		String name;
 		String description;
-		int release;
+		int api_level;
 	};
 
 	Vector<Device> devices;
@@ -1495,7 +1495,7 @@ void EditorExportPlatformAndroid::_device_poll_thread(void *ud) {
 						if (ea->devices[j].id == ldevices[i]) {
 							d.description = ea->devices[j].description;
 							d.name = ea->devices[j].name;
-							d.release = ea->devices[j].release;
+							d.api_level = ea->devices[j].api_level;
 						}
 					}
 
@@ -1516,7 +1516,7 @@ void EditorExportPlatformAndroid::_device_poll_thread(void *ud) {
 						String vendor;
 						String device;
 						d.description + "Device ID: " + d.id + "\n";
-						d.release = 0;
+						d.api_level = 0;
 						for (int j = 0; j < props.size(); j++) {
 
 							String p = props[j];
@@ -1527,9 +1527,9 @@ void EditorExportPlatformAndroid::_device_poll_thread(void *ud) {
 							} else if (p.begins_with("ro.build.display.id=")) {
 								d.description += "Build: " + p.get_slice("=", 1).strip_edges() + "\n";
 							} else if (p.begins_with("ro.build.version.release=")) {
-								const String release_str = p.get_slice("=", 1).strip_edges();
-								d.description += "Release: " + release_str + "\n";
-								d.release = release_str.to_int();
+								d.description += "Release: " + p.get_slice("=", 1).strip_edges() + "\n";
+							} else if (p.begins_with("ro.build.version.sdk=")) {
+								d.api_level = p.get_slice("=", 1).to_int();
 							} else if (p.begins_with("ro.product.cpu.abi=")) {
 								d.description += "CPU: " + p.get_slice("=", 1).strip_edges() + "\n";
 							} else if (p.begins_with("ro.product.manufacturer=")) {
@@ -1688,7 +1688,7 @@ Error EditorExportPlatformAndroid::run(int p_device, int p_flags) {
 	args.push_back("shell");
 	args.push_back("am");
 	args.push_back("start");
-	if (bool(EDITOR_DEF("android/force_system_user", false)) && devices[p_device].release >= 17) { // Multi-user introduced in Android 17
+	if (bool(EDITOR_DEF("android/force_system_user", false)) && devices[p_device].api_level >= 17) { // Multi-user introduced in Android 17
 		args.push_back("--user");
 		args.push_back("0");
 	}

--- a/platform/android/export/export.cpp
+++ b/platform/android/export/export.cpp
@@ -1596,7 +1596,11 @@ Error EditorExportPlatformAndroid::run(int p_device, int p_flags) {
 	//export_temp
 	ep.step("Exporting APK", 0);
 
-	p_flags |= EXPORT_REMOTE_DEBUG_LOCALHOST; // Needed for adb reverse
+	const bool use_remote = (p_flags & EXPORT_REMOTE_DEBUG) || (p_flags & EXPORT_DUMB_CLIENT);
+	const bool use_reverse = devices[p_device].api_level >= 21;
+
+	if (use_reverse)
+		p_flags |= EXPORT_REMOTE_DEBUG_LOCALHOST;
 
 	String export_to = EditorSettings::get_singleton()->get_settings_path() + "/tmp/tmpexport.apk";
 	Error err = export_project(export_to, true, p_flags);
@@ -1645,40 +1649,54 @@ Error EditorExportPlatformAndroid::run(int p_device, int p_flags) {
 		return ERR_CANT_CREATE;
 	}
 
-	args.clear();
-	args.push_back("-s");
-	args.push_back(devices[p_device].id);
-	args.push_back("reverse");
-	args.push_back("--remove-all");
-	err = OS::get_singleton()->execute(adb, args, true, NULL, NULL, &rv);
+	if (use_remote) {
+		if (use_reverse) {
 
-	if (p_flags & EXPORT_REMOTE_DEBUG) {
+			static const char *const msg = "** Device API >= 21; debugging over USB **";
+			EditorNode::get_singleton()->get_log()->add_message(msg);
+			print_line(String(msg).to_upper());
 
-		int dbg_port = (int)EditorSettings::get_singleton()->get("network/debug_port");
-		args.clear();
-		args.push_back("-s");
-		args.push_back(devices[p_device].id);
-		args.push_back("reverse");
-		args.push_back("tcp:" + itos(dbg_port));
-		args.push_back("tcp:" + itos(dbg_port));
+			args.clear();
+			args.push_back("-s");
+			args.push_back(devices[p_device].id);
+			args.push_back("reverse");
+			args.push_back("--remove-all");
+			err = OS::get_singleton()->execute(adb, args, true, NULL, NULL, &rv);
 
-		err = OS::get_singleton()->execute(adb, args, true, NULL, NULL, &rv);
-		print_line("Reverse result: " + itos(rv));
-	}
+			if (p_flags & EXPORT_REMOTE_DEBUG) {
 
-	if (p_flags & EXPORT_DUMB_CLIENT) {
+				int dbg_port = (int)EditorSettings::get_singleton()->get("network/debug_port");
+				args.clear();
+				args.push_back("-s");
+				args.push_back(devices[p_device].id);
+				args.push_back("reverse");
+				args.push_back("tcp:" + itos(dbg_port));
+				args.push_back("tcp:" + itos(dbg_port));
 
-		int fs_port = EditorSettings::get_singleton()->get("file_server/port");
+				err = OS::get_singleton()->execute(adb, args, true, NULL, NULL, &rv);
+				print_line("Reverse result: " + itos(rv));
+			}
 
-		args.clear();
-		args.push_back("-s");
-		args.push_back(devices[p_device].id);
-		args.push_back("reverse");
-		args.push_back("tcp:" + itos(fs_port));
-		args.push_back("tcp:" + itos(fs_port));
+			if (p_flags & EXPORT_DUMB_CLIENT) {
 
-		err = OS::get_singleton()->execute(adb, args, true, NULL, NULL, &rv);
-		print_line("Reverse result2: " + itos(rv));
+				int fs_port = EditorSettings::get_singleton()->get("file_server/port");
+
+				args.clear();
+				args.push_back("-s");
+				args.push_back(devices[p_device].id);
+				args.push_back("reverse");
+				args.push_back("tcp:" + itos(fs_port));
+				args.push_back("tcp:" + itos(fs_port));
+
+				err = OS::get_singleton()->execute(adb, args, true, NULL, NULL, &rv);
+				print_line("Reverse result2: " + itos(rv));
+			}
+		} else {
+
+			static const char *const msg = "** Device API < 21; debugging over Wi-Fi **";
+			EditorNode::get_singleton()->get_log()->add_message(msg);
+			print_line(String(msg).to_upper());
+		}
 	}
 
 	ep.step("Running on Device..", 3);

